### PR TITLE
enable gocritic, unconvert and unused linters

### PR DESCRIPTION
**Pull request by mmorel-35**
Signed-off-by: Matthieu MOREL <matthieu.morel35@gmail.com>

### CODE REVIEW

File: template/template.go [hunk 1]
```diff
@@ -422,7 +422,7 @@ func (te Expander) ExpandHTML(templateFiles []string) (result string, resultErr
 		}
 	}()
 
-	tmpl := html_template.New(te.name).Funcs(html_template.FuncMap(te.funcMap))
+	tmpl := html_template.New(te.name).Funcs(te.funcMap)
```
Comment (mmorel-35) [inline template/template.go:425]:
```suggestion
	tmpl := html_template.New(te.name).Funcs(html_template.FuncMap(te.funcMap))
```

File: discovery/vultr/vultr.go [hunk 2]
```diff
@@ -202,10 +202,9 @@ func (d *Discovery) listInstances(ctx context.Context) ([]govultr.Instance, erro
 
 		if meta.Links.Next == "" {
 			break
-		} else {
-			listOptions.Cursor = meta.Links.Next
-			continue
 		}
+		listOptions.Cursor = meta.Links.Next
+		continue
```
Comment (bboreham) [inline discovery/vultr/vultr.go:207]: This `continue` is redundant, right?

File: promql/engine.go [hunk 3]
```diff
@@ -741,8 +741,7 @@ func subqueryTimes(path []parser.Node) (time.Duration, time.Duration, *int64) {
 		ts                    int64 = math.MaxInt64
 	)
 	for _, node := range path {
-		switch n := node.(type) {
-		case *parser.SubqueryExpr:
+		if n, ok := node.(*parser.SubqueryExpr); ok {
```
Comment (bboreham) [inline promql/engine.go:744]: This one seems slightly worse to me, in that the existing code allows for new types to be added.

File: promql/engine.go [hunk 4]
```diff
@@ -778,8 +776,7 @@ func (ng *Engine) findMinMaxTime(s *parser.EvalStmt) (int64, int64) {
 				maxTimestamp = end
 			}
 			evalRange = 0
-
-		case *parser.MatrixSelector:
+		} else if n, ok := node.(*parser.MatrixSelector); ok {
```
Comment (bboreham) [inline promql/engine.go:779]: This one seems definitely worse.  We're taking out `else if` in other places.

File: template/template.go [hunk 5]
```diff
@@ -421,7 +421,7 @@ func (te Expander) ExpandHTML(templateFiles []string) (result string, resultErr
 			}
 		}
 	}()
-
+	//nolint:unconvert // This ensure compatibility with previous go version
```
Comment (bboreham) [inline template/template.go:424]: Can you add a little more explanation here, e.g. some hint at what the error would be.

File: tsdb/compact_test.go [hunk 6]
```diff
@@ -1452,12 +1452,12 @@ func TestSparseHistogramSpaceSavings(t *testing.T) {
...
-		//{1000, 100, 1, 0},
-		//{1000, 15, 3, 5},
-		//{1000, 50, 3, 3},
-		//{1000, 100, 3, 2},
+		// {1000, 15, 1, 0},
+		// {1000, 50, 1, 0},
+		// {1000, 100, 1, 0},
+		// {1000, 15, 3, 5},
+		// {1000, 50, 3, 3},
+		// {1000, 100, 3, 2},
```
Comment (bboreham) [inline tsdb/compact_test.go:1460]: Nit: I would rather delete these lines than create diffs with a space.

**Review by bboreham — COMMENTED** [review]
Thanks for your efforts.  Some of the changes are worthwhile; I'm not bothered about most of them, but I do object to the ones replacing a type switch with if-else.

File: discovery/vultr/vultr.go (hunk 2 above)
Comment (mmorel-35) [reply discovery/vultr/vultr.go]: Yes !

File: template/template.go (hunk 5 above)
Comment (mmorel-35) [reply template/template.go]: I changed the description consequently

**Comment by beorn7** [comment]
FTR: https://github.com/prometheus/prometheus/pull/12252/commits/5b53aa11083233963bd477d19134492355d9ed0f is the commit that removes almost all `} else if` occurrences, with a detailed reasoning in the commit description. As said there, I don't think we need to enforce removal af `} else if`, but we should definitely not enforce its presence for `switch` statements with two branches. (Which this PR takes into account, so LGTM.)

**Comment by beorn7** [comment]
Sorry for the merge conflicts. They are caused by merging #12252 . I guess they should be very easy to resolve.

**Review by roidelapluie — APPROVED** [review]
I am merging this to avoid further conflicts. Why did you add skip-cache? can you remove this in a further commit?

**Pull request merged**